src/extension/index.ts

```diff
@@ -1,15 +1,41 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { ExtensionContext } from 'vscode';
-import * as path from 'path';
+import { Event, EventEmitter, ExtensionContext, notebooks, extensions, NotebookEditor } from 'vscode';
+import { OpenImageInPlotViewer, SaveImageAs, IsJupyterExtensionInstalled } from './constants';
+
+export async function activate(
+    context: ExtensionContext
+): Promise<{
+    onDidReceiveMessage: Event<{ editor: NotebookEditor; message: OpenImageInPlotViewer | SaveImageAs }>;
+}> {
+    const onDidReceiveMessage = new EventEmitter<{
+        editor: NotebookEditor;
+        message: OpenImageInPlotViewer | SaveImageAs;
+    }>();
+    const messaging = notebooks.createRendererMessaging('jupyter-notebook-renderer');
+    context.subscriptions.push(
+        messaging.onDidReceiveMessage(({ editor, message }) => {
+            const msg = message as OpenImageInPlotViewer | SaveImageAs | IsJupyterExtensionInstalled;
+            if (!msg.type) {
+                return;
+            }
+            if (msg.type === 'isJupyterExtensionInstalled') {
+                const isJupyterExtAvailable = extensions.getExtension('ms-toolsai.jupyter');
+                void messaging.postMessage(
+                    <IsJupyterExtensionInstalled>{
+                        type: 'isJupyterExtensionInstalled',
+                        response: isJupyterExtAvailable
+                    },
+                    editor
+                );
+                return;
+            }
+            onDidReceiveMessage.fire({ editor, message: msg });
```

Comment:
I don't want to expose the editor, notebook Document is sufficient, but `NotebookEditor.NotebookDocument` property is only available in proposed API, means this extension will need proposed API.
Hence returning the editor object to the other side so the other extension (Jupyter) can extract the document from that object.

Need to avoid proposed API, as this extnsion can be used in github.dev and other places

Reply:
Could you just use the URI of the document? Then the listener can use that to look it up?

Reply:
Yes we can, but that means we need to perform an unnecessary lookup (iteration of the notebooks).